docs: update op secure integration overview (#1592)
Layr-Labs/eigenda: docs/spec/src/integration/spec/7-op-secure-integration-workflow.md

Reviewer: Just realized this file doesn't go here. It should go under the rollup-stack section.
![image](https://github.com/user-attachments/assets/3bc929a3-e3b8-4d74-b2f2-b17fddbcf898)

Can you make this a `4.2.1 - OP Secure Integration` page?

This document presents an overview on how EigenDA plugs into Optimism (OP) Stack.
- `write` and `read` path in a L2 rollup
- Why the `read` path must stay live (even with a misbehaving op-batcher)
- Adding an EigenDA stage to the OP derivation pipeline
- Hokulea, Rust library that defines and implements the Eigenda derivation rule
- How Hokulea works in both interactive fault-proof VMs and zkVMs

## Write and Read path in L2 consensus

A rollup system can be split into two parts: write path to L1 and read path from L1

| Path      | Direction | Purpose                                    | Main actor                   |
| --------- | --------- | ------------------------------------------ | ---------------------------- |
| **Write**| L2 → L1   | Low cost L2 block production with user transactions  | `op-batcher` + EigenDA proxy |
| **Write**| Direct on L1   | Censorship resistance + Deposit        | Rollup users        + Opimism Portal     |
| **Read**  | L1 → L2   | Safety – all nodes see the same block list | OP derivation pipeline       |

- The `write path` ensures the liveness of the L2 consensus. It consists of L2 batches produced by op-batcher and L1  deposit transactions.

> - The `read path` controls the safety of the L2 consensus. It ensures that all L2 consensus nodes see an identical list of L2 batchrs and L1 sytem and deposits transactions, such that an EVM can produce identical L2 state

Reviewer: ```suggestion
- The `read path` controls the safety of the L2 consensus. It ensures that all L2 consensus nodes see an identical list of L2 batches and L1 sytem and deposits transactions, such that an EVM can produce identical L2 state
```

# EigenDA OP Secure integration

> This document presents an overview on how EigenDAA plugs into Optimism (OP) Stack.

Reviewer: ```suggestion
This document presents an overview on how EigenDA plugs into Optimism (OP) Stack.
```

- `write` and `read` path in a L2 rollup
- Why the `read` path must stay live (even with a misbehaving op-batcher)
- Adding an EigenDA stage to the OP derivation pipeline

> - Hokulea, Rust library that defines and implement the Eigenda derivation rule

Reviewer: ```suggestion
- Hokulea, Rust library that defines and implements the Eigenda derivation rule
```

- How Hokulea works in both interactive fault-proof VMs and zkVMs

## Write and Read path in L2 consensus

> A rollup system can be splitted into two parts: write path to L1 and read path from L1

Reviewer: ```suggestion
A rollup system can be split into two parts: write path to L1 and read path from L1
```

| Path      | Direction | Purpose                                    | Main actor                   |
| --------- | --------- | ------------------------------------------ | ---------------------------- |

> | **Write**| L2 → L1   | Liveness – keep blocks flowing             | `op-batcher` + EigenDA proxy |

Reviewer: not 100% accurate. blocks keeps flowing regardless of whether l2 tx batches are submitted or not.

Author: that is not most of the traffic, 

also, op-node wait for sequencing_window to close, which takes 12h

Author: de2e24b1bfff836afdb25dea17487e999511afc7

Author: so it will stall I think [E, E+SWS] https://specs.optimism.io/protocol/derivation.html#overview

Reviewer: sure but liveness is not the right term then, since liveness just means "eventually a block gets produced"

| **Write**| Direct on L1   | Censorship resistance + Deposit        | Rollup users        + Opimism Portal     |
| **Read**  | L1 → L2   | Safety – all nodes see the same block list | OP derivation pipeline       |

> - The `write path` ensures the liveness of the L2 consensus. It consists of L2 blocks produced by op-batch and L1 induced transactions.

Reviewer: ```suggestion
- The `write path` ensures the liveness of the L2 consensus. It consists of L2 batches produced by op-batcher and L1  deposit transactions.
```

> - The `read path` controls the safety of the L2 consensus. It ensures all L2 consensus node sees an identical list of L2 blocks, then a EVM can produce identical L2 state

Reviewer: is this true? feel like l1 deposit txs are also needed for safety because without this you can't withdraw your money

Author: oh, I can add l1 tx

Author: d71a88fcc803ac99970437c779449db87bcef7a4

Reviewer: why do you say "induced". ive never seen that used anywhere. also its batches not blocks. blocks = l1 deposit txs + batches + system txs

> - The `read path` controls the safety of the L2 consensus. It ensures all L2 consensus node sees an identical list of L2 blocks, then a EVM can produce identical L2 state

Reviewer: ```suggestion
- The `read path` controls the safety of the L2 consensus. It ensures that all L2 consensus nodes see an identical list of L2 blocks, such that an EVM can produce identical L2 state
```

Reviewer: this is not 100% accurate since DP produces payload attributes and the EVM produces the blocks. but maybe not worth going into that level of details

> If the read path stalls, honest nodes can’t reach the block height needed to dispute a bad state root. As the result so it must never stall.

Reviewer: ```suggestion
If the read path stalls, honest nodes can’t reach the block height needed to dispute a bad state root.
```

Reviewer: It *could* stall though right?

### L2 Write path (happy-flow)
- op-batcher bundles user txs.

> - Sends batch to EigenDA proxy, which converts it to an Eigenda blob. Proxy sends blob to EigenDA, and forwards the returned certificate to op-batcher

Reviewer: ```suggestion
- Sends compressed batches to EigenDA proxy, which converts them into an Eigenda blob. Proxy sends blob to EigenDA, and forwards the returned certificate to op-batcher
```

> - Post certificate are posted to the L1 Rollup Inbox.

Reviewer: ```suggestion
- EigenDA certificates are posted to the L1 Rollup Inbox.
```

Both Eigenda proxy and Hokulea implement the eigenda blob derivation.

## L2 Read path with EigenDA

As in the diagram, op-nodes use the `read-path` on the eigenda-proxy to fetch EigenDA blob. The proxy checks
- certificate has sufficient stake and is valid
- certificate is not stale
- retrieve & KZG-checked blob from EigenDA
- Decode and pass the data onward

More information can be found in the page [secure integration](./6-secure-integration.md). The key properties which EigenDA derivation strives to keep are

- Determinism – one unique blob per DA certificate.
- Liveness – discard anything that could halt the chain.

Both eigenda-proxy and hokulea hold those properties.

## Proving correctness on L1

The security of rollup is determined by if there are provable ways to challenge incorrect L2 state posted on L1.
In this section, we discuss our OP secure integraton library **Hokulea**.

### Short intro to OP FPVM

The correctness of a L2 state is determined by the derivation rules, which are implemented in both Go [op-program](https://github.com/ethereum-optimism/optimism/tree/develop/op-program) and Rust [Kona](https://github.com/op-rs/kona/tree/main).

> With interactive fault proof, the derivation logics are packaged into a binary ELF file, which can be run inside a FPVM (Cannon is the implementation).

Reviewer: ```suggestion
With interactive fault proof, the derivation logics are packaged into a binary ELF file, which can be run inside a FPVM (Cannon, Asterisc, etc.).
```

The FPVM requires both the ELF binary and data (L2 batches and L1 deposits) to be able to rerun the derivation pipeline.

> The ida is to repeat what op-node has done to reach consensus, except that in FPVM, every execution is traceable and challengable.

Reviewer: ```suggestion
The idea is to repeat what op-node has done to reach consensus, except that in FPVM, every execution is traceable and challengeable.
```

Data is provided to FPVM in the form of preimage oracle. OP spec has defined rules such that all data in the preimage oracle are verifiable on L1.

> A FPVM creates a binding system on the relation between (software logic, data, VM opcode, and output).

Reviewer: ```suggestion
```

### Hokulea

> Hokulea builds on top of OP Kona derivation pipeline to integrate EigenDA as a Data Availability Source. Hokulea provides traits, implementation

Reviewer: Is it really building on the top? To me its more of a plugin (build on "side") system that augment's kona's DP.

Reviewer: also to me mentioning traits is like saying that we use "consts" and "functions". We arent providing traits like kona that are meant to be implemented by external 3rd party providers, so don't think we should mention this

Author: I don't prefer plugin, as if it is something you can do hot-swap

uses traits exposed by

Author: de2e24b1bfff836afdb25dea17487e999511afc7

for EigenDA part of derivation pipeline, such that those logics can be compiled into ELF together with Kona.

Hokulea also extends preimage oracle for EigenDA, which is able to provide the verifiable interface for answering

> - if a DA cert is correct

Reviewer: ```suggestion
- whether a DA cert is correct
```

> - what is the current recency window to determine if a blob is stale

Reviewer: ```suggestion
- what is the current recency window to determine if a cert is stale
```

More information about the communication spec is at [Hokulea](https://github.com/Layr-Labs/hokulea/tree/master/docs). Both extension to preimage
oracle and derivation logics allows for 

> - deterministically derive rollup payload from a EigenDA certificate

Reviewer: ```suggestion
- deterministically deriving a rollup payload from an EigenDA certificate
```

> - discard DA certs that can stall the derivation pipeline

Reviewer: ```suggestion
- discarding DA certs that can stall the derivation pipeline
```

### Canoe

> We developed a rust library called [**Canoe**](https://github.com/Layr-Labs/hokulea/tree/master/canoe#1protocol-overview) that uses zk validity proof to provide ways 
> - to efficiently verify the cert validity via verifying a zk validity proof
> - to verify validity of a DA cert within zkVM, this enables zkVM integration

Reviewer: ```suggestion
We developed a rust library called [**Canoe**](https://github.com/Layr-Labs/hokulea/tree/master/canoe#1protocol-overview) that uses zk validity proof to efficiently verify the cert validity on L1 or inside a zkEVM.
```

Reviewer: is it any zkvm or just zkevm?

Author: Both can, sp1 and risc0 use zkvm to run evm. Themselves are not zkevm

Author: I wil just say zkvm

Author: de2e24b1bfff836afdb25dea17487e999511afc7

### Hokulea integration with zkVM

> Unlike interactively challenge game with fault proof, a zk proof has a property that only the honest party can create a valid zk proof with respect to

Reviewer: ```suggestion
Unlike interactive challenge game with fault proof, a zk proof has a property that only the honest party can create a valid zk proof with respect to
```

the correct derivation rule.

> Hence,a n incorrect party can raise a challenge but is unable to defend its position.

Reviewer: ```suggestion
Hence, a malicious party can raise a challenge but is unable to defend its position.
```

> - The Hokulea+Kona derivation are compiled into ELF

Reviewer: ```suggestion
- The Hokulea+Kona derivation is compiled into ELF for the needed environment (RiscV zkEVM or one of the FPVMs)
```

Author: zkVM

Author: de2e24b1bfff836afdb25dea17487e999511afc7

- The Hokulea+Kona preimage oracle are prepared, where the validity of DA cert is provided by Canoe

> - zkVM takes preimage and verifys it, then feed the data into the ELF containing the derivation logics

Reviewer: ```suggestion
- zkVM takes preimage and verifies it, then feeds the data into the ELF containing the derivation logics
```

- zkVM produces a proof about the execution

Hokulea is currently integrating with [OP-succinct](https://github.com/succinctlabs/op-succinct) and [OP-Kailua](https://github.com/risc0/kailua).
For an integration guide, please refer to the [preloader](https://github.com/Layr-Labs/hokulea/tree/master/example/preloader) example for zk integration.

### Rust Kzg Bn254 library

> The constraint also requires all eigenda blob with respect to the kzg commitment in the DA cert. We developed a similar library to `c-kzg`s called

Reviewer: ```suggestion
The constraint also requires all eigenda blob with respect to the kzg commitment in the DA cert. We developed a similar library to `c-kzg` called
```
